src/GraphQL/Validation/DocumentValidator.cs

```diff
@@ -81,42 +81,48 @@ public class DocumentValidator : IDocumentValidator
-                Schema = schema,
-                Document = document,
-                TypeInfo = new TypeInfo(schema),
-                UserContext = userContext,
-                Inputs = inputs
-            };
+                context.OriginalQuery = originalQuery ?? document.OriginalQuery;
+                context.Schema = schema;
+                context.Document = document;
+                context.TypeInfo = new TypeInfo(schema);
+                context.UserContext = userContext;
+                context.Inputs = inputs;
 
-            List<INodeVisitor> visitors;
+                List<INodeVisitor> visitors;
 
-            if (useOnlyStandardRules)
-            {
-                // No async/await related allocations since all standard rules return completed tasks from ValidateAsync.
-                visitors = new List<INodeVisitor>();
-                foreach (var rule in (List<IValidationRule>)rules) // no iterator boxing
+                if (useOnlyStandardRules)
                 {
-                    var visitor = rule.ValidateAsync(context)?.Result;
-                    if (visitor != null)
-                        visitors.Add(visitor);
+                    // No async/await related allocations since all standard rules return completed tasks from ValidateAsync.
+                    visitors = new List<INodeVisitor>();
+                    foreach (var rule in (List<IValidationRule>)rules) // no iterator boxing
+                    {
+                        var visitor = rule.ValidateAsync(context)?.Result;
+                        if (visitor != null)
+                            visitors.Add(visitor);
+                    }
+                }
+                else
+                {
+                    var awaitedVisitors = rules.Select(x => x.ValidateAsync(context)).Where(x => x != null);
+                    visitors = (await Task.WhenAll(awaitedVisitors)).ToList();
                 }
-            }
-            else
-            {
-                var awaitedVisitors = rules.Select(x => x.ValidateAsync(context)).Where(x => x != null);
-                visitors = (await Task.WhenAll(awaitedVisitors)).ToList();
-            }   
 
-            visitors.Insert(0, context.TypeInfo);
+                visitors.Insert(0, context.TypeInfo);
 
-            new BasicVisitor(visitors).Visit(document, context);
+                new BasicVisitor(visitors).Visit(document, context);
 
-            return context.HasErrors
-                ? new ValidationResult(context.Errors)
-                : (IValidationResult)SuccessfullyValidatedResult.Instance;
+                return context.HasErrors
+                    ? new ValidationResult(context.Errors)
+                    : (IValidationResult)SuccessfullyValidatedResult.Instance;
+            }
+            finally
+            {
+                context.Reset();
+                System.Threading.Interlocked.CompareExchange(ref _reusableValidationContext, context, null);
```

Comment:
```suggestion
                if (success)
                {
                    context.Reset();
                    System.Threading.Interlocked.CompareExchange(ref _reusableValidationContext, context, null);
                }
```

Reply:
The validation context is accessible by third-party code.  Just like field resolvers, these changes will not re-use a context where the validation results in an error, just in case someone stores the context temporarily so information contained therein can be saved to the database after validation is complete.

Reply:
done